test/simulator/main/org/apache/cassandra/simulator/paxos/HistoryValidator.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.simulator.paxos;
+
+import javax.annotation.Nullable;
+
+public interface HistoryValidator
+{
+    Checker start(Observation observation);
+
+    void log(@Nullable Integer pk);
+
+    interface Checker extends AutoCloseable
+    {
+        void witnessRead(int pk, int count, int[] seq);
+        void witnessWrite(int pk);
```

Comment:
I'm not 100% keen on this, as it disguises the fact that the `id` is what we're logging here. I think perhaps(?) I would rather accept the `id` redundantly and verify it is the same, or some other change I haven't thought of yet.

Reply:
Not a fan as that is captured by `log(Observation), so would allow users to provide the wrong thing and make usage more verbose.

Reply:
They couldn't provide the wrong thing if you validate it?

As it stands, it is unclear what it is we are logging as written, which wasn't the case before. It's implicit knowledge that it is the observation id we are writing, which is especially obtuse since the id exists whether or not we write. 

I'd be inclined then to just have `LinearizabilityValidator` implement an API similar to the one provided by the StrictSerializabilityVerifier, and introduce a common interface that we call directly, as it was IMO clearer what was being reported.

Reply:
> As it stands, it is unclear what it is we are logging as written, which wasn't the case before. It's implicit knowledge that it is the observation id we are writing, which is especially obtuse since the id exists whether or not we write.

I understand your point, my biggest issue is the verbosity.  If we add the id then I question why `log` should see `Observation`?  Maybe the following is better?

```
public interface HistoryValidator {
  Checker log(int start, int end);
  void log(@Nullable Integer pk);

  interface Checker extends AutoCloseable {
    void witnessRead(int pk, int count, int[] sequence);
    void witnessWrite(int pk, int id);
  }
}
```

The main issue with this interface is that `LinearizabilityValidator` doesn't see `Observation` so would need a refactor in `HistoryChecker`; not a large one but would be needed (issue is in `org.apache.cassandra.simulator.paxos.HistoryChecker#witness`), we did speak about dropping `LinearizabilityValidator`, maybe this is better motivation?

I personally prefer the existing interface as you can't mess up and put event id in pk and pk in event id, but feel the above interface is a compromise?  Thoughts?

Reply:
yeah, that works for me. Though looking at it written there, the two different `log` methods probably aren't ideal. Perhaps rename the prior `log` method to `print` to match the underlying impls?

Reply:
pushed, but had to make slight changes

```
void witnessRead(int pk, int id, int count, int[] seq);
void witnessWrite(int pk, int id, boolean success);
```

id is needed in reader, and we need to know if the write was success or not.

Reply:
Splitting hairs here, so feel free to ignore, but I think it might be confusing to a reader that the antonym of success in this context is unknown. Instead of accepting a `success` boolean parameter, it might be nice to have `witnessWriteSuccess` and `witnessWriteTimeout`. 

Also, another brief suggestion to accept/discard: perhaps the method to create the `Checker` should be `witness` and we can call it `Witness witness = checker.witness()` and invoke `witness.read(...)` and `witness.writeSuccess(...)`? `witness.writeTimeout(...)`?

Reply:
pushed.  I like `witness`, I do feel its nicer.  I think `write` is still better in the current usage, but cool with default methods...

I prefer `checker.write(pk, outcome.id, outcome.isSuccess());` in the current use than

```
if (outcome.isSuccess()) checker.writeSuccess(pk, outcome.id);
else checker.writeUnknownFailure(pk, outcome.id);
```

I also prefer `UnknownFailure` than `Timeout` as it includes timeout + other issues